src/cell_list.h

```diff
@@ -0,0 +1,226 @@
+        int calc_distance() const {
+            int distance = max_distance;
+            if( min_bounds.x >= 0 ) {
+                distance = std::min( distance, origin.x - min_bounds.x );
+            }
+            if( min_bounds.y >= 0 ) {
+                distance = std::min( distance, origin.y - min_bounds.y );
+            }
+            if( max_bounds.x < MAXX ) {
+                distance = std::min( distance, max_bounds.x - origin.x );
+            }
+            if( max_bounds.y < MAXY ) {
+                distance = std::min( distance, max_bounds.y - origin.y );
+            }
+            return distance;
+        }
+        void add( const std::vector<std::pair<point, T>> &cells ) {
+            for( const std::pair<point, T> &candidate : cells ) {
+                add( candidate );
+            }
+        }
+        void add( const std::pair<point, T> &candidate ) {
+            const int distance = rl_dist( candidate.first, origin );
+            if( distance > max_distance ) {
+                return;
+            }
+            sorted_elements[distance].push_back( candidate );
+        }
+        void load() {
+            constexpr point rank_offset{ CELL_SIZE, CELL_SIZE };
+            min_bounds -= rank_offset;
+            max_bounds += rank_offset;
+            distance_to_nearest_cell_wall = calc_distance();
+            cell_ranks_loaded++;
+            const point load_offset( cell_ranks_loaded, cell_ranks_loaded );
+            const point min_cell = origin_cell - load_offset;
+            const point max_cell = origin_cell + load_offset;
+            const bool min_x_inbounds = min_cell.x >= 0;
+            const bool min_y_inbounds = min_cell.y >= 0;
+            const bool max_x_inbounds = max_cell.x < list->num_cells_x();
+            const bool max_y_inbounds = max_cell.y < list->num_cells_x();
+            // Draw a box...
+            for( int x = min_cell.x; x <= max_cell.x; ++x ) {
+                if( x < 0 || x >= list->num_cells_x() ) {
+                    continue;
+                }
+                if( min_y_inbounds ) {
+                    add( list->cells[x][min_cell.y] );
+                }
+                if( max_y_inbounds ) {
+                    add( list->cells[x][max_cell.y] );
+                }
+            }
+            for( int y = min_cell.y + 1; y <= max_cell.y - 1; ++y ) {
+                if( y < 0 || y >= list->num_cells_y() ) {
+                    continue;
+                }
+                if( min_x_inbounds ) {
+                    add( list->cells[min_cell.x][y] );
+                }
+                if( max_x_inbounds ) {
+                    add( list->cells[max_cell.x][y] );
+                }
+            }
+        }
+        const cell_list<T, CELL_SIZE, MAXX, MAXY> *list;
+        const point origin;
+        const point origin_cell; // = origin / CELL_SIZE
+        const int max_distance;
+        std::vector<std::vector<std::pair<point, T>>> sorted_elements;
```

Comment:
Copying whole pairs from `cell_list` can be somewhat expensive (especially assuming `T` is something non-trivial). Maybe store iterators / pointers instead? However, this implies immutability of `list` during traversal, but I suppose it's a prerequisite anyways.

Reply:
Yes this can easily capture a pointer instead of make a copy.